packages/core/src/model.ts

```diff
@@ -0,0 +1,44 @@
+import { assign } from './actions';
+import { AssignAction, Assigner, EventObject, PropertyAssigner } from './types';
+import { mapValues } from './utils';
+
+export interface ContextModel<TC, TE extends EventObject> {
+  context: TC | (() => TC);
+  actions: {
+    [key: string]: Assigner<TC, TE> | PropertyAssigner<TC, TE>;
+  };
+  withAssigners: <
+    T extends {
+      [key: string]: Assigner<TC, TE> | PropertyAssigner<TC, TE>;
+    }
+  >(
+    assigners: T
+  ) => ContextModel<TC, TE> & {
+    actions: {
+      [K in keyof T]: T[K] extends Assigner<TC, infer E>
+        ? AssignAction<TC, E>
+        : T[K] extends PropertyAssigner<TC, infer E>
+        ? AssignAction<TC, E>
+        : never;
+    };
+  };
+}
+
+export function createModel<TContext, TEvent extends EventObject>(
+  initialState: TContext | (() => TContext)
```

Comment:
Lazy initialization is definitely useful when you only want something to be created:
- per machine (one instance per invoked machine)
- only when the machine is created.

But also, this is future-proofing for v5, where `spawn` will be provided _internally_ (and properly typed) rather than externally, so you can imagine an API like this:

```js
initialContext: ({ spawn }) => {
  return {
    someActorRef: spawn(/* ... */)
  }
}
```